framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java

```diff
@@ -4631,6 +4619,452 @@ private void makeGroundTargetType(
+   * type variables in capturedType are the same object when they are refer to the same type
+   * variable.
+   *
+   * <p>To use, Call {@code #copy(AnnotatedDeclaredType, AnnotatedDeclaredType)} rather than a visit
+   * method.
+   */
+  private final NonWildcardTypeArgCopier captureTypeArgCopier = new NonWildcardTypeArgCopier();
+
+  /**
+   * Copy the non-wildcard type args from {@code uncapturedType} to {@code captureType}. Also,
+   * ensure that type variables in {@code capturedType} are the same object when they refer to the
+   * same type variable.
+   *
+   * <p>To use, Call {@code #copy(AnnotatedDeclaredType, AnnotatedDeclaredType)} rather than a visit
+   * method.
+   */
+  private class NonWildcardTypeArgCopier extends AnnotatedTypeCopier {
+
+    /**
+     * Copy the non-wildcard type args from {@code uncapturedType} to {@code capturedType}. Also,
+     * ensure that type variables {@code capturedType} are the same object when they are refer to
+     * the same type variable.
+     *
+     * @param uncapturedType a declared type that has not under gone capture conversion
+     * @param capturedType the captured version of {@code uncapturedType} before it has been
+     *     annotated.
+     */
+    private void copy(AnnotatedDeclaredType uncapturedType, AnnotatedDeclaredType capturedType) {
+
+      IdentityHashMap<AnnotatedTypeMirror, AnnotatedTypeMirror> originalToCopy =
+          new IdentityHashMap<>();
+      AnnotatedDeclaredType original = uncapturedType;
+      originalToCopy.put(original, capturedType);
+      int size = uncapturedType.getTypeArguments().size();
+
+      AnnotatedTypeMirror[] newTypeArgs = new AnnotatedTypeMirror[size];
+      Map<TypeVariable, AnnotatedTypeMirror> typeVarMap = new HashMap<>(size);
+      // Copy the non-wildcard type args from uncapturedType to newTypeArgs.
+      // If the non-wildcard type arg is type var add it to typeVarMap.
+      for (int i = 0; i < size; i++) {
+        AnnotatedTypeMirror uncapturedArg = uncapturedType.getTypeArguments().get(i);
+        if (uncapturedArg.getKind() != TypeKind.WILDCARD) {
+          AnnotatedTypeMirror copyArg = visit(uncapturedArg, originalToCopy);
+          if (copyArg.getKind() == TypeKind.TYPEVAR) {
+            typeVarMap.put(((AnnotatedTypeVariable) copyArg).getUnderlyingType(), copyArg);
+          }
+          newTypeArgs[i] = copyArg;
+        }
+      }
+
+      // Substitute the type variables in each type argument of capturedType using typeVarMap
+      // created above.
+      // This makes type variables in capturedType the same object when they are the same type
+      // variable.
+      for (int i = 0; i < size; i++) {
+        AnnotatedTypeMirror uncapturedArg = uncapturedType.getTypeArguments().get(i);
+        AnnotatedTypeMirror capturedArg = capturedType.getTypeArguments().get(i);
+        if (uncapturedArg.getKind() == TypeKind.WILDCARD) {
+          // Note: this if statement can't be replaced with if
```

Comment:
Please put comments above the code that they refer to.  (That is the standard convention.)